Work around OpenCV highgui bug

roehling opened the pull request:
If OpenCV is compiled against GTK, running the GUI code in a dedicated thread will trigger a crash. While this is ultimately something for the OpenCV people to fix, we can work around the problem in the `image_view` node and fake a GUI thread with the help of  `ros::Timer`.

As a bonus, this PR also fixes the mutex assertion that occurs when the `image_view` program tries to shut down.

Fixes #201

roehling committed:
Work around an OpenCV bug with GTK and threading

roehling committed:
While we're at it, work around the mutex assertion failure on exit

JWhitleyWork commented:
Tested as working and fixing the mutex issue.

JWhitleyWork commented:
Can @vrabaud please review this? This is breaking image_view in Melodic.

flixr commented:
You should use a `SteadyTimer` instead of a normal Timer (which could even pause on sim time).

roehling committed:
Switch to SteadyTimer as suggested in review

roehling committed:
Use WallTimer for backwards compatibility with ROS Indigo

roehling committed:
Force dpkg update to work around Travis CI problem

roehling committed:
Forgot to change variable type

knorth55 commented:
What is the status of this issue?
I'm suffering the same problem in #201 and fix it with this PR.

JWhitleyWork commented:
The status is that the maintainers in ros-perception appear to have totally abandoned maintenance of ROS1 while working on ROS2. I have tried to bring some attention to the issues and PRs on this group of repositories (see https://groups.google.com/forum/m/#!topic/ros-sig-perception/pTA4tz2yEZU and https://discourse.ros.org/t/perception-sig-maintainers-still-maintaining/7065) but no one is answering. I'm not sure what else to do.

knorth55 commented:
@JWhitleyAStuff Thank you for the infomation. I also posted on ROSDiscourse for reviewing this PR.

nitheeshkl commented:
@roehling Thanks for the fix! Meanwhile, since the nodelet also suffers from the same issue, it would be great if you include the same changes for nodelet as well. (instead of someone raising another issue & PR for the image nodelet in the future.)

roehling committed:
Refresh GUI on image update as well

roehling committed:
Also fix image_view nodelet

I had to remove the GTK workaround, since it creates symbol collisions
between GTK2 and GTK3.

roehling committed:
Pass --include-eol-distros to enable Jade build

JWhitleyWork commented:
@nitheeshkl - Can you re-review now that @roehling has added the suggested changes?

nitheeshkl reviewed (approved):
Looks good to me. Verified this PR resolves the OpenCV crash issue at runtime.

nitheeshkl commented:
@JWhitleyAStuff reviewed and verified on Melodic with Ubuntu 18.04 env. If only any of the maintainers can merge this PR now, would save a lot of people from wasting time on this bug!

knorth55 commented:
This PR is still waiting for reviewer response as described in
https://discourse.ros.org/t/perception-sig-maintainers-still-maintaining/7065.
I still want this PR to be merged because `image_view` is a fundamental feature.

Maintenance is a big problem, and  I also want to inform you ROS orphaned package project.
https://discourse.ros.org/t/ros-orphaned-package-maintainer-info-rep-152-for-review/7906

JWhitleyWork reviewed (approved):
LGTM and tested as working.
